### PR TITLE
[review] Dask performance tuning + Dask plugin + Smart Cache handling + Composite Node performance tuning + Template Node

### PR DESCRIPTION
**Pull request by yidong72**
In this PR, I made following changes:
1. By default, the Dask dataframe meta data is using the meta_setup.
2. Added the necessary persist node. The changes of 1-2 make existing dask workflow 2 times faster.
3. Use the ml.xgboos for XGBoost strategies. 
4. Added the SimpleParallelNode to make parallel computation easy.
5. Added a notebook as an example use of SimpleParallelNode
6. Improves the DaskComputeNode to handle multiple ports simultaneously. 
7. Move the DaskComputeNode, SimpleParallelNode and PersistNode into a separate Greenflow plugin
8. Change the core plots to matplotlib as bqplot is not stable enough. 

I tested all the notebooks.

### CODE REVIEW

**Comment by yidong72** [comment]
I added the doc string and tested all the notebooks. It is working and ready for review!

**Comment by yidong72** [comment]
This should be my last round of change.

The composite node only build the graph once in the `update` function so `ports_setup`, `meta_setup`, `conf_schema` can be called cheaply. 

`GridRandomSearchNode` is also fixed to removed the strong requirement that all upstream graphs need to be serializable.

**Review by avolkov1 — COMMENTED** [review]
I'm still reviewing and testing. Left two comments so far.

File: gQuant/plugins/dask_plugin/greenflow_dask_plugin/simpleParallelNode.py [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+            for i in range(len(self.conf['map'])):
+                val = iternum
+                map_obj = self.conf['map'][i]
+                xpath = map_obj['xpath']
+                expr = parse(xpath)
+                expr.update(replaceObj, val)
+
+    def _process(self, inputs, iternum):
+        """
+        Composite computation
```
[the remark this answers is not in the document]
Comment (yidong72) [reply gQuant/plugins/dask_plugin/greenflow_dask_plugin/simpleParallelNode.py]: Done

**Review by avolkov1 — COMMENTED** [review]
I left a few comments. The logic in the `SimpleNodeMixin` seems to me like it really should be part of `NodeTaskGraphMixin`. Let me know what you think.

Can you explain a bit more regarding the `SimpleNodeMixin`, especially its `update`, `ports_setup` and `meta_setup` methods, if you disagree with me regarding that belonging in the `NodeTaskGraphMixin`? I think all nodes should have this logic injected when they are within a task graph. It shouldn't be necessary to inherit explicitly from any mixin for a Node's implementation.

File: greenflow/greenflow/dataframe_flow/simpleNodeMixin.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+        module_cache[type_str] = clsobj
+        return clsobj
+
+    def meta_setup(self, required={}):
+        if hasattr(self, 'meta_data_cache'):
+            return self.meta_data_cache
+        if hasattr(self, 'input_meta'):
+            input_meta = self.input_meta
+        else:
+            input_meta = self.get_input_meta()
```
[the remark this answers is not in the document]
Comment (avolkov1) [reply greenflow/greenflow/dataframe_flow/simpleNodeMixin.py:359]: This comment was incorrect given an incomplete understanding of how `SimpleNodeMixin` worked.
PR https://github.com/yidong72/gQuant/pull/4 has the refactoring that I would prefer.

File: greenflow/greenflow/dataframe_flow/simpleNodeMixin.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+        assert isinstance(variable, str)
+        e = re.search('^\${(.*):(.*)}$', variable) # noqa
+        if e is None and variable.startswith('$'):
+            raise ValueError("varaible format is wrong")
+        if e is None:
+            return None
+        groups = e.groups()
+        return groups
+
+    def ports_setup(self):
```
[the remark this answers is not in the document]
Comment (avolkov1) [reply greenflow/greenflow/dataframe_flow/simpleNodeMixin.py:269]: This comment was incorrect given an incomplete understanding of how `SimpleNodeMixin` worked.
PR https://github.com/yidong72/gQuant/pull/4 has the refactoring that I would prefer.

File: greenflow/greenflow/dataframe_flow/simpleNodeMixin.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+
+    def update(self):
+        """
+        This function is called after the computation graph is
+        constructed by Taskgraph.build method. It assumes the
+        graph structure is up to date
+
+        If need information from self.get_connected_inports()
+        or self.get_input_meta() to modify the ports_setup or
+        meta_setup, this is the place to do it.
```
[the remark this answers is not in the document]
Comment (avolkov1) [reply greenflow/greenflow/dataframe_flow/simpleNodeMixin.py:131]: In the suggested PR https://github.com/yidong72/gQuant/pull/4 I incorporated the caching logic into `NodeTaskGraphExtensionMixin`.

File: greenflow/greenflow/dataframe_flow/simpleNodeMixin.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+        self.port_outports = {}
+        self.meta_inports = {}
+        self.meta_outports = {}
+        self.META_OP_DELETION = 'deletion'
+        self.META_OP_ADDITION = 'addition'
+        self.META_OP_RETENTION = 'retention'
+        self.META_REF_INPUT = 'input'
+        self.META_OP = 'meta_op'
+        self.META_DATA = 'data'
+        self.META_ORDER = 'order'
```
[the remark this answers is not in the document]
Comment (avolkov1) [reply greenflow/greenflow/dataframe_flow/simpleNodeMixin.py:113]: The refactoring of `MetaDataSchema` is done in PR https://github.com/yidong72/gQuant/pull/4.

**Comment by yidong72** [comment]
I made a few changes to  make the template state private. I checked all the notebooks and unit tests.

**Review by avolkov1 — APPROVED** [review]
Looks good. Everything works. I'll merge this.

**Pull request merged**